File: bazelrunner/src/main/java/org/jetbrains/bsp/bazel/bazelrunner/BazelRunnerBuilder.java
package org.jetbrains.bsp.bazel.bazelrunner;

import ch.epfl.scala.bsp4j.BuildTargetIdentifier;
import com.google.common.collect.ImmutableList;
import java.util.ArrayList;
import java.util.List;
import org.jetbrains.bsp.bazel.bazelrunner.params.BazelQueryKindParameters;
import org.jetbrains.bsp.bazel.bazelrunner.utils.BazelArgumentsUtils;

public class BazelRunnerBuilder {

  private final BazelRunner bazelRunner;
  private final String bazelCommand;

  private final List<String> flags;
  private final List<String> arguments;

  BazelRunnerBuilder(BazelRunner bazelRunner, String bazelCommand) {
    this.bazelRunner = bazelRunner;
    this.bazelCommand = bazelCommand;
    this.flags = new ArrayList<>();
    this.arguments = new ArrayList<>();
  }

  public BazelRunnerBuilder withFlag(String bazelFlag) {
    flags.add(bazelFlag);

    return this;
  }

  public BazelRunnerBuilder withFlags(List<String> bazelFlags) {
    flags.addAll(bazelFlags);

    return this;
  }

  public BazelRunnerBuilder withArgument(String bazelArgument) {
    arguments.add(bazelArgument);

    return this;
  }

  public BazelRunnerBuilder withArguments(List<String> bazelArguments) {
    arguments.addAll(bazelArguments);

    return this;
  }

  public BazelRunnerBuilder withTargets(List<String> bazelTargets) {
    String joinedTargets = BazelArgumentsUtils.getJoinedBazelTargets(bazelTargets);
    arguments.add(joinedTargets);

    return this;
  }

  public BazelRunnerBuilder withTargets(
      List<BuildTargetIdentifier> includedTargets, List<BuildTargetIdentifier> excludedTargets) {
    var joinedTargets = BazelArgumentsUtils.joinBazelTargets(includedTargets, excludedTargets);
    arguments.add(joinedTargets);

    return this;
  }

  public BazelRunnerBuilder withMnemonic(List<String> bazelTargets, List<String> languageIds) {
    String argument = BazelArgumentsUtils.getMnemonicWithJoinedTargets(bazelTargets, languageIds);
    arguments.add(argument);

    return this;
  }

  public BazelRunnerBuilder withKind(BazelQueryKindParameters bazelParameter) {
    return withKinds(ImmutableList.of(bazelParameter));
  }

  public BazelRunnerBuilder withKinds(List<BazelQueryKindParameters> bazelParameters) {
    String argument = BazelArgumentsUtils.getQueryKindForPatternsAndExpressions(bazelParameters);
    arguments.add(argument);

    return this;
  }

  public BazelRunnerBuilder withKindsAndExcept(
      BazelQueryKindParameters parameters, String exception) {
    String argument =
        BazelArgumentsUtils.getQueryKindForPatternsAndExpressionsWithException(
            ImmutableList.of(parameters), exception);
    arguments.add(argument);

    return this;
  }

  public BazelProcess executeBazelCommand() {
    return bazelRunner.runBazelCommand(bazelCommand, flags, arguments);
  }

  public BazelProcess executeBazelBesCommand() {
    return bazelRunner.runBazelCommandBes(bazelCommand, flags, arguments);
  }
}

File: bazelrunner/src/main/kotlin/org/jetbrains/bsp/bazel/bazelrunner/BazelRunnerBuilder.kt
package org.jetbrains.bsp.bazel.bazelrunner

import ch.epfl.scala.bsp4j.BuildTargetIdentifier
import com.google.common.collect.ImmutableList
import org.jetbrains.bsp.bazel.bazelrunner.params.BazelQueryKindParameters
import org.jetbrains.bsp.bazel.bazelrunner.params.BazelFlag
import org.jetbrains.bsp.bazel.bazelrunner.utils.BazelArgumentsUtils
import org.jetbrains.bsp.bazel.workspacecontext.TargetsSpec
import java.nio.file.Path

open class BazelRunnerBuilder internal constructor(
    private val bazelRunner: BazelRunner,
    private val bazelCommand: List<String>,
) {
    private val globalFlags = listOf<String>(BazelFlag.toolTag())
    private val flags = globalFlags.toMutableList()
    private val arguments = mutableListOf<String>()
    private val environmentVariables = mutableMapOf<String, String>()
    private var useBuildFlags = false

    fun withUseBuildFlags(useBuildFlags: Boolean = true): BazelRunnerBuilder {
        this.useBuildFlags = useBuildFlags
        return this
    }

    fun withFlag(bazelFlag: String): BazelRunnerBuilder {
        flags.add(bazelFlag)
        return this
    }

    fun withFlags(bazelFlags: List<String>?): BazelRunnerBuilder {
        flags.addAll(bazelFlags.orEmpty())
        return this
    }

    fun withArgument(bazelArgument: String): BazelRunnerBuilder {
        arguments.add(bazelArgument)
        return this
    }

    fun withArguments(bazelArguments: List<String>?): BazelRunnerBuilder {
        arguments.addAll(bazelArguments.orEmpty())
        return this
    }

    open fun withTargets(bazelTargets: List<String>): BazelRunnerBuilder {
        val joinedTargets = BazelArgumentsUtils.getJoinedBazelTargets(bazelTargets)
        arguments.add(joinedTargets)
        return this
    }

    open fun withTargets(targetsSpec: TargetsSpec): BazelRunnerBuilder {
        val joinedTargets = BazelArgumentsUtils.joinBazelTargets(targetsSpec.values, targetsSpec.excludedValues)
        arguments.add(joinedTargets)
        return this
    }

    open fun withTargets(
        includedTargets: List<BuildTargetIdentifier>,
        excludedTargets: List<BuildTargetIdentifier>
    ): BazelRunnerBuilder? {
        val joinedTargets = BazelArgumentsUtils.joinBazelTargets(includedTargets, excludedTargets)
        arguments.add(joinedTargets)
        return this
    }

    fun withMnemonic(bazelTargets: List<String>, languageIds: List<String>): BazelRunnerBuilder {
        val argument = BazelArgumentsUtils.getMnemonicWithJoinedTargets(bazelTargets, languageIds)
        arguments.add(argument)
        return this
    }

    fun withKind(bazelParameter: BazelQueryKindParameters): BazelRunnerBuilder {
        return withKinds(ImmutableList.of(bazelParameter))
    }

    fun withKinds(bazelParameters: List<BazelQueryKindParameters>): BazelRunnerBuilder {
        val argument = BazelArgumentsUtils.getQueryKindForPatternsAndExpressions(bazelParameters)
        arguments.add(argument)
        return this
    }

    fun withKindsAndExcept(
        parameters: BazelQueryKindParameters,
        exception: String
    ): BazelRunnerBuilder {
        val argument = BazelArgumentsUtils.getQueryKindForPatternsAndExpressionsWithException(
            listOf(parameters), exception
        )
        arguments.add(argument)
        return this
    }

    fun withEnvironment(environment: List<Pair<String, String>>): BazelRunnerBuilder {
        environmentVariables.putAll(environment)
        return this
    }

    fun executeBazelCommand(originId: String? = null, parseProcessOutput: Boolean = true, serverPid: Long? = null): BazelProcess {
        return bazelRunner.runBazelCommand(
            bazelCommand,
            flags,
            arguments,
            environmentVariables,
            originId,
            parseProcessOutput,
            useBuildFlags,
            serverPid,
        )
    }

    fun executeBazelBesCommand(originId: String? = null, buildEventFile: Path, serverPid: Long? = null): BazelProcess {
        return bazelRunner.runBazelCommandBes(
            bazelCommand,
            flags,
            arguments,
            environmentVariables,
            originId,
            buildEventFile.toAbsolutePath(),
            serverPid,
        )
    }
}
